### PR TITLE
[APB-3939]: (Refactor) Add Endpoint to Get Invitations by Identifier

### PR DESCRIPTION
**Pull request by JTY009**

### CODE REVIEW

File: app/uk/gov/hmrc/agentclientauthorisation/controllers/ClientInvitationsController.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+  private def determineService(service: String, identifier: String): ClientService =
+    service match {
+      case "MTDITID" if MtdItIdType.isValid(identifier) =>
+        ClientService(Service.MtdIt, MtdItIdType.createUnderlying(identifier))
+      case "NI" if NinoType.isValid(identifier) =>
+        ClientService(Service.PersonalIncomeRecord, NinoType.createUnderlying(identifier))
+      case "VRN" if VrnType.isValid(identifier) => ClientService(Service.Vat, VrnType.createUnderlying(identifier))
+      case "UTR" if UtrType.isValid(identifier) => ClientService(Service.Trust, UtrType.createUnderlying(identifier))
+      case _                                    => throw new Exception(s"Unsupported $service and identifier $identifier")
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentclientauthorisation/controllers/ClientInvitationsController.scala]: this should ideally trigger bad_request as the service is invalid (with a message saying Unsupported $service)

File: app/uk/gov/hmrc/agentclientauthorisation/controllers/ClientInvitationsController.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  ecp: Provider[ExecutionContextExecutor],
+  @Named("old.auth.stride.enrolment") oldStrideRole: String,
+  @Named("new.auth.stride.enrolment") newStrideRole: String)
+    extends BaseClientInvitationsController(invitationsService, metrics, authConnector, auditService) {
+
+  implicit val ec: ExecutionContext = ecp.get
+
+  private val strideRoles = Seq(oldStrideRole, newStrideRole)
+
+  private def determineService(service: String, identifier: String): ClientService =
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentclientauthorisation/controllers/ClientInvitationsController.scala]: can we return a tuple here which makes more meaningful instead of a case class ? because `ClientService` doesn't really tell what it does

File: app/uk/gov/hmrc/agentclientauthorisation/controllers/ClientInvitationsController.scala (hunk 1 above)
Comment (JTY009) [reply app/uk/gov/hmrc/agentclientauthorisation/controllers/ClientInvitationsController.scala]: I think I can rearrange it

File: app/uk/gov/hmrc/agentclientauthorisation/controllers/ClientInvitationsController.scala [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      implicit request => implicit currentUser =>
+        implicit val authTaxId: Option[ClientIdentifier[TaxIdentifier]] =
+          if (currentUser.credentials.providerType == "GovernmentGateway")
+            Some(ClientIdentifier(clientService.clientIdentifier))
+          else None
+        getInvitations(clientService.service, ClientIdentifier(clientService.clientIdentifier), status)
+    }
+  }
+
+  override protected def agencyLink(invitation: Invitation): Option[String] = None
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentclientauthorisation/controllers/ClientInvitationsController.scala]: why is this always None?

File: app/uk/gov/hmrc/agentclientauthorisation/controllers/ClientInvitationsHal.scala [hunk 4]
```diff
@@ -38,6 +39,9 @@ trait ClientInvitationsHal {
         routes.NiClientInvitationsController.getInvitation(Nino(invitation.clientId.value), invitation.invitationId)
       case Service.Vat =>
         routes.VatClientInvitationsController.getInvitation(Vrn(invitation.clientId.value), invitation.invitationId)
+      case Service.Trust =>
+        //TODO To Be Added for APB-3938
+        Call("GET", s"/clients/UTR/:utr/invitations/received/${invitation.invitationId.value}")
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentclientauthorisation/controllers/ClientInvitationsHal.scala]: can we make this as above links `routes.VatClientInvitationsController......` ? 
we dont need to implement it in this PR, just an endpoint is enough

File: app/uk/gov/hmrc/agentclientauthorisation/controllers/ClientInvitationsHal.scala [hunk 5]
```diff
@@ -77,8 +87,10 @@ trait ClientInvitationsHal {
           routes.MtdItClientInvitationsController.getInvitation(MtdItId(i.clientId.value), i.invitationId)
         case Service.PersonalIncomeRecord =>
           routes.NiClientInvitationsController.getInvitation(Nino(i.clientId.value), i.invitationId)
-        case Service.Vat => routes.VatClientInvitationsController.getInvitation(Vrn(i.clientId.value), i.invitationId)
-
+        case Service.Vat   => routes.VatClientInvitationsController.getInvitation(Vrn(i.clientId.value), i.invitationId)
+        case Service.Trust =>
+          //TODO To Be Added for APB-3938
+          Call("GET", s"/clients/UTR/:utr/invitations/received/${i.invitationId.value}")
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentclientauthorisation/controllers/ClientInvitationsHal.scala]: same as above

File: it/uk/gov/hmrc/agentclientauthorisation/controllers/ClientAcceptInvitationsControllerISpec.scala [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+import uk.gov.hmrc.agentclientauthorisation.model.ClientIdentifier.ClientId
+import uk.gov.hmrc.agentclientauthorisation.model.{Invitation, Service}
+import uk.gov.hmrc.agentclientauthorisation.repository.{InvitationsRepository, InvitationsRepositoryImpl}
+import uk.gov.hmrc.agentmtdidentifiers.model._
+import uk.gov.hmrc.domain.Nino
+
+import scala.concurrent.Future
+import scala.concurrent.ExecutionContext.Implicits.global
+
+class ClientAcceptInvitationsControllerISpec extends BaseISpec {
```
Comment (sureshhmrc) [inline it/uk/gov/hmrc/agentclientauthorisation/controllers/ClientAcceptInvitationsControllerISpec.scala]: is this relevant to current PR ?

File: it/uk/gov/hmrc/agentclientauthorisation/controllers/ClientGetInvitationsByIdentifierControllerISpec.scala [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+import uk.gov.hmrc.agentclientauthorisation.model.ClientIdentifier.ClientId
+import uk.gov.hmrc.agentclientauthorisation.model.{Invitation, Service}
+import uk.gov.hmrc.agentclientauthorisation.repository.{InvitationsRepository, InvitationsRepositoryImpl}
+import uk.gov.hmrc.agentmtdidentifiers.model.{Arn, MtdItId, Utr, Vrn}
+import uk.gov.hmrc.domain.Nino
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+class ClientGetInvitationsByIdentifierControllerISpec extends BaseISpec {
```
Comment (sureshhmrc) [inline it/uk/gov/hmrc/agentclientauthorisation/controllers/ClientGetInvitationsByIdentifierControllerISpec.scala]: controller is `ClientInvitationsController` so can we rename this spec accordingly  `ClientInvitationsControllerISpec`?

File: it/uk/gov/hmrc/agentclientauthorisation/controllers/ClientGetInvitationsByIdentifierControllerISpec.scala [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+      "found VAT Invitations" in new StubsSuccessfulScenariosTest(personal, Service.Vat) {
+        val result = await(controller.getInvitations("VRN", vrn.value, None)(request))
+
+        status(result) shouldBe 200
+
+        ((jsonBodyOf(result) \ "_embedded" \ "invitations")(0) \ "arn").as[String] shouldBe arn2.value
+        ((jsonBodyOf(result) \ "_embedded" \ "invitations")(1) \ "arn").as[String] shouldBe arn.value
+      }
+
+      "found Trust Invitations" in new StubsSuccessfulScenariosTest(personal, Service.Trust) {
```
Comment (sureshhmrc) [inline it/uk/gov/hmrc/agentclientauthorisation/controllers/ClientGetInvitationsByIdentifierControllerISpec.scala]: all the above 4 tests can be managed with a single test if we iterate trough the available services ?

File: it/uk/gov/hmrc/agentclientauthorisation/controllers/ClientRejectInvitationsControllerISpec.scala [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+import uk.gov.hmrc.agentclientauthorisation.model.ClientIdentifier.ClientId
+import uk.gov.hmrc.agentclientauthorisation.model.{Invitation, Service}
+import uk.gov.hmrc.agentclientauthorisation.repository.{InvitationsRepository, InvitationsRepositoryImpl}
+import uk.gov.hmrc.agentmtdidentifiers.model._
+import uk.gov.hmrc.domain.Nino
+
+import scala.concurrent.Future
+import scala.concurrent.ExecutionContext.Implicits.global
+
+class ClientRejectInvitationsControllerISpec extends BaseISpec {
```
Comment (sureshhmrc) [inline it/uk/gov/hmrc/agentclientauthorisation/controllers/ClientRejectInvitationsControllerISpec.scala]: is this relevant to current PR ?  or can we postpone until next PR when you do PUTs ?

File: app/uk/gov/hmrc/agentclientauthorisation/connectors/AuthActions.scala [hunk 10]
```diff
@@ -118,19 +122,34 @@ class AuthActions @Inject()(metrics: Metrics, val authConnector: AuthConnector)
       .contains(clientId)
   }
 
-  def AuthorisedClientOrStrideUser[T](clientId: TaxIdentifier, strideRoles: Seq[String])(body: RequestAndCurrentUser)(
-    implicit ec: ExecutionContext): Action[AnyContent] =
+  def determineService(service: String, identifier: String)(implicit hc: HeaderCarrier): Either[Future[Result], (Service, TaxIdentifier)] =
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentclientauthorisation/connectors/AuthActions.scala]: this can simply be `Either[BadRequest, (Service, TaxIdentifier)]`

File: app/uk/gov/hmrc/agentclientauthorisation/connectors/AuthActions.scala [hunk 11]
```diff
@@ -118,19 +122,34 @@ class AuthActions @Inject()(metrics: Metrics, val authConnector: AuthConnector)
...
+            case Right((clientService, clientId)) =>
+              creds.providerType match {
+                case "GovernmentGateway" if hasRequiredEnrolmentMatchingIdentifier(enrolments, clientId) =>
+                  body(request)(CurrentUser(enrolments, creds, clientService, clientId))
+                case "PrivilegedApplication" if hasRequiredStrideRole(enrolments, strideRoles) =>
+                  body(request)(CurrentUser(enrolments, creds, clientService, clientId))
+                case _ =>
+                  Future successful NoPermissionToPerformOperation
+              }
+            case Left(error) => error
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentclientauthorisation/connectors/AuthActions.scala]: if you change Either as above then this will change to `case Left(error) => Future.successfule(error)`

File: app/uk/gov/hmrc/agentclientauthorisation/controllers/ClientInvitationsHal.scala (hunk 4 above)
Comment (sureshhmrc) [reply app/uk/gov/hmrc/agentclientauthorisation/controllers/ClientInvitationsHal.scala]: did you miss this or you wanted to keep it like this ?

File: app/uk/gov/hmrc/agentclientauthorisation/controllers/ClientInvitationsHal.scala (hunk 5 above)
Comment (sureshhmrc) [reply app/uk/gov/hmrc/agentclientauthorisation/controllers/ClientInvitationsHal.scala]: same as above ?

File: app/uk/gov/hmrc/agentclientauthorisation/model/Service.scala [hunk 12]
```diff
@@ -123,3 +123,5 @@ object ClientIdentifier {
 
   implicit def wrap[T <: TaxIdentifier](taxId: T): ClientIdentifier[T] = ClientIdentifier(taxId)
 }
+
+case class ClientService(service: Service, clientIdentifier: TaxIdentifier)
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentclientauthorisation/model/Service.scala]: do we still need this ?
